### ISSUE TITLE
Analyzer for multiple enumerations of Enumerables

### ISSUE THREAD

**Issue by mikadumont** [dotnet/roslyn-analyzers#3946]
Direct customer ask copying here:

>Showing warnings for multiple enumerations of Enumerables, this actually helped us with a critical bug fix recently.  
![image](https://user-images.githubusercontent.com/46729679/89211367-902dcf80-d576-11ea-96c2-551727b3fe6a.png)

**Comment by mavasani** [comment]
I was thinking about this analyzer over the weekend :-). I am going to give this a shot this week.

**Comment by Evangelink** [comment]
Dupe of #3133. @mavasani which one do you prefer to close?

**Comment by drewnoakes** [comment]
A challenging case for this analyzer:

```c#
public int M(IEnumerable<int> e)
{
    Requires.NotNull(e, nameof(e));

    return e.First();
}
```

How can you know that `Requires.NotNull` doesn't enumerate the object yet `Enumerable.First` does?

**Comment by nschuessler** [comment]
This is a common case we see, somewhat related:

```csharp
class Something
{
     public void One()
     {
          List<string> l = new List<string>()
          {
                 "One", "Two", "Three"
           };

           this.Two(l);
     }

     private void Two(IEnumerable<string> l)
     {
            int count = l.Count();

           this.Three(l);
     }

     private void Three(IEnumerable<string> l)
     {
         foreach (string s in l)
         {
             Console.WriteLine(s);
         }
     }
}
```

Better to pass `IReadOnlyList<string>`  or `ICollection<string>` when passed between methods.

**Comment by drewnoakes** [comment]
@nschuessler I would expect an analyzer to fire in the `Two` method from your example. What won't be clear to the analyzer however is whether the method that receives the enumerable actually enumerates it (i.e. it may just be validating it is not null).

**Comment by mavasani** [comment]
@Evangelink We are considering this analyzer for .NET6 release, would you be interested in implementing it?

**Comment by Evangelink** [comment]
I am a bit short in time lately so I'd say "yes but I can't really commit on ETA". So if anyone else wants to pick this, fine by me.

**Comment by rui-ktei** [comment]
Google took me here. This is actually a much needed feature. In our team, we use Resharper which does underline "possible multiple enumeration" issue, but the problem is that people still ignore it sometimes, so I was searching a way to force this type of "suggestion" to be a compilation error to force devs to pay attention.

I remember there's a way to configure roslyn-analyzer (through .editconfig?) to set certain code compliance to different level. But then I realized that roslyn-analyzer itself doesn't support "multiple enumeration" detection yet.

I know it's probably going to be included in .net6, but in the meantime, do you guys have any suggestion regarding how I should force devs to avoid such issue rather than purely relying on Resharper suggestion?

Many thanks.

**Comment by Cosifne** [comment]
> Google took me here. This is actually a much needed feature. In our team, we use Resharper which does underline "possible multiple enumeration" issue, but the problem is that people still ignore it sometimes, so I was searching a way to force this type of "suggestion" to be a compilation error to force devs to pay attention.
> 
> I remember there's a way to configure roslyn-analyzer (through .editconfig?) to set certain code compliance to different level. But then I realized that roslyn-analyzer itself doesn't support "multiple enumeration" detection yet.
> 
> I know it's probably going to be included in .net6, but in the meantime, do you guys have any suggestion regarding how I should force devs to avoid such issue rather than purely relying on Resharper suggestion?
> 
> Many thanks.

Tag @mavasani for visibility

**Comment by Evangelink** [comment]
I would be careful with this analyzer because I have seen a couple of codebases with a lot of "false-positives" where basically most methods were taking `IEnumerable<T>` but a `List<T>` (or something else) was _ALWAYS_ passed so although the issue is theoretically correct there was actually no real performance benefit.

Also, one of the suggested fixes by Resharper was often used blindly by dev:

```csharp
void M(IEnumerable<XXX> collection)
{
   var list = collection as IList<XXX> ?? collection.ToList();
}
```

**Comment by drewnoakes** [comment]
My experience using ReSharper is that this analysis found many performance issues over the years. Methods that assume enumeration is"free" may be safe today, but are prone to introduce non obvious performance or correctness issues later.

I would not suggest adding a codefix. There is no straightforward way to fix multiple enumeration automatically.

If a codebase uses materialised enumerables extensively and finds such a diagnostic noisy, they can turn it off.
